File: labs/08-devops/devops-solution/ContosoShopEasy/Services/PaymentService.cs
using ContosoShopEasy.Models;
using ContosoShopEasy.Data;

namespace ContosoShopEasy.Services
{
    public class PaymentService
    {
        // Security vulnerability: Hardcoded configuration values (but won't trigger GitHub Secret Scanning)
        private const string PAYMENT_GATEWAY_URL = "https://api.contoso-payments.com";
        private const string MERCHANT_NAME = "ContosoShopEasy";
        private const string GATEWAY_VERSION = "v2.1";

        private readonly OrderRepository _orderRepository;

        public PaymentService(OrderRepository orderRepository)
        {
            _orderRepository = orderRepository;
        }

        // Payment processing method
        public bool ProcessPayment(string cardNumber, string cardHolderName, string expiryDate, string cvv, decimal amount)
        {
            string maskedCardNumber = MaskCardNumber(cardNumber);

            Console.WriteLine($"[DEBUG] Processing payment for card ending in {maskedCardNumber}");
            Console.WriteLine($"[DEBUG] Card holder: {cardHolderName}");
            Console.WriteLine($"[DEBUG] Amount: ${amount}");

            // Security vulnerability: Log configuration details
            Console.WriteLine($"[DEBUG] Using payment gateway: {PAYMENT_GATEWAY_URL}");
            Console.WriteLine($"[DEBUG] Merchant: {MERCHANT_NAME}");
            Console.WriteLine($"[DEBUG] Gateway version: {GATEWAY_VERSION}");

            // Simulate payment validation (vulnerable)
            if (!ValidateCardNumber(cardNumber))
            {
                Console.WriteLine($"[ERROR] Invalid card number ending in {maskedCardNumber}");
                return false;
            }

            if (!ValidateExpiryDate(expiryDate))
            {
                Console.WriteLine($"[ERROR] Invalid or expired date: {expiryDate}");
                return false;
            }

            // Simulate payment processing
            Console.WriteLine("[INFO] Connecting to payment gateway...");
            Thread.Sleep(1000); // Simulate network delay

            // Security vulnerability: Generate predictable transaction IDs
            string transactionId = GenerateTransactionId(cardNumber, amount);

            var paymentInfo = new PaymentInfo
            {
                Method = PaymentMethod.CreditCard,
                CardLastFourDigits = maskedCardNumber,
                CardType = DetectCardType(cardNumber),
                CardHolderName = cardHolderName,
                ExpiryDate = expiryDate,
                Amount = amount,
                ProcessedDate = DateTime.UtcNow,
                Status = PaymentStatus.Approved,
                TransactionId = transactionId
            };

            Console.WriteLine($"[SUCCESS] Payment processed successfully!");
            Console.WriteLine($"[DEBUG] Transaction ID: {transactionId}");

            Console.WriteLine($"[LOG] Payment completed - Card: ****{paymentInfo.CardLastFourDigits}, Amount: ${amount}, Transaction: {transactionId}");

            return true;
        }

        // Returns only the last 4 digits of the card number for safe display and logging
        private static string MaskCardNumber(string cardNumber)
        {
            string digitsOnly = new string((cardNumber ?? string.Empty).Where(char.IsDigit).ToArray());
            return digitsOnly.Length >= 4 ? digitsOnly[^4..] : digitsOnly;
        }

        // Detects the card brand from the Issuer Identification Number range
        private static string DetectCardType(string cardNumber)
        {
            string digitsOnly = new string((cardNumber ?? string.Empty).Where(char.IsDigit).ToArray());

            if (digitsOnly.StartsWith("4"))
                return "Visa";

            if (digitsOnly.Length >= 2 && int.TryParse(digitsOnly.Substring(0, 2), out int firstTwoDigits) && firstTwoDigits >= 51 && firstTwoDigits <= 55)
                return "Mastercard";

            if (digitsOnly.StartsWith("34") || digitsOnly.StartsWith("37"))
                return "American Express";

            if (digitsOnly.StartsWith("6011") || digitsOnly.StartsWith("65"))
                return "Discover";

            return "Unknown";
        }

        // Vulnerable card validation
        private bool ValidateCardNumber(string cardNumber)
        {
            // Security vulnerability: Weak validation - only checks length
            if (string.IsNullOrEmpty(cardNumber))
                return false;

            // Remove spaces and dashes
            cardNumber = cardNumber.Replace(" ", "").Replace("-", "");

            // Security vulnerability: Accept any 13-19 digit number
            return cardNumber.Length >= 13 && cardNumber.Length <= 19 && cardNumber.All(char.IsDigit);
        }

        private bool ValidateExpiryDate(string expiryDate)
        {
            // Security vulnerability: Basic validation only
            if (string.IsNullOrEmpty(expiryDate) || !expiryDate.Contains("/"))
                return false;

            var parts = expiryDate.Split('/');
            if (parts.Length != 2)
                return false;

            if (int.TryParse(parts[0], out int month) && int.TryParse(parts[1], out int year))
            {
                if (year < 100) year += 2000; // Convert YY to YYYY
                var expiry = new DateTime(year, month, 1).AddMonths(1).AddDays(-1);
                return expiry >= DateTime.Now;
            }

            return false;
        }

        // Security vulnerability: Predictable transaction ID generation
        private string GenerateTransactionId(string cardNumber, decimal amount)
        {
            // Vulnerable: Using predictable pattern
            string lastFour = cardNumber.Length >= 4 ? cardNumber.Substring(cardNumber.Length - 4) : cardNumber;
            string timestamp = DateTime.Now.ToString("yyyyMMddHHmm");
            string amountStr = amount.ToString("F2").Replace(".", "");
            
            return $"TXN_{timestamp}_{lastFour}_{amountStr}";
        }

        public bool RefundPayment(string transactionId, decimal amount)
        {
            // Security vulnerability: Log refund details
            Console.WriteLine($"[DEBUG] Processing refund for transaction: {transactionId}, Amount: ${amount}");
            Console.WriteLine($"[DEBUG] Using payment gateway: {PAYMENT_GATEWAY_URL}");

            // Simulate refund processing
            Console.WriteLine("[INFO] Processing refund...");
            Thread.Sleep(500);

            Console.WriteLine($"[SUCCESS] Refund processed for transaction: {transactionId}");
            return true;
        }

        // Method to get payment history - with security issues
        public List<PaymentInfo> GetPaymentHistory(int userId)
        {
            Console.WriteLine($"[DEBUG] Retrieving payment history for user: {userId}");
            
            // In a real app, this would query the database
            // For demo purposes, we'll return empty list
            return new List<PaymentInfo>();
        }
    }
}

File: labs/08-devops/devops-solution-py/contoso_shop_easy/services/payment_service.py
import time
from calendar import monthrange
from datetime import datetime, timezone
from decimal import Decimal

from ..data.order_repository import OrderRepository
from ..models.order import PaymentInfo, PaymentMethod, PaymentStatus


class PaymentService:
    # Security vulnerability: Hardcoded configuration values (but won't trigger GitHub Secret Scanning)
    PAYMENT_GATEWAY_URL = "https://api.contoso-payments.com"
    MERCHANT_NAME = "ContosoShopEasy"
    GATEWAY_VERSION = "v2.1"

    def __init__(self, order_repository: OrderRepository) -> None:
        self._order_repository = order_repository

    # Payment processing method
    def process_payment(self, card_number: str, card_holder_name: str, expiry_date: str, cvv: str, amount: Decimal) -> bool:
        card_last_four_digits = self._mask_card_number(card_number)

        print(f"[DEBUG] Processing payment for card ending in {card_last_four_digits}")
        print(f"[DEBUG] Card holder: {card_holder_name}")
        print(f"[DEBUG] Amount: ${amount}")

        # Security vulnerability: Log configuration details
        print(f"[DEBUG] Using payment gateway: {self.PAYMENT_GATEWAY_URL}")
        print(f"[DEBUG] Merchant: {self.MERCHANT_NAME}")
        print(f"[DEBUG] Gateway version: {self.GATEWAY_VERSION}")

        # Simulate payment validation (vulnerable)
        if not self._validate_card_number(card_number):
            print(f"[ERROR] Invalid card number ending in {card_last_four_digits}")
            return False

        if not self._validate_expiry_date(expiry_date):
            print(f"[ERROR] Invalid or expired date: {expiry_date}")
            return False

        # Simulate payment processing
        print("[INFO] Connecting to payment gateway...")
        time.sleep(1)  # Simulate network delay

        # Security vulnerability: Generate predictable transaction IDs
        transaction_id = self._generate_transaction_id(card_number, amount)

        payment_info = PaymentInfo(
            method=PaymentMethod.CREDIT_CARD,
            card_last_four_digits=card_last_four_digits,
            card_type=self._detect_card_type(card_number),
            card_holder_name=card_holder_name,
            expiry_date=expiry_date,
            amount=amount,
            processed_date=datetime.now(timezone.utc),
            status=PaymentStatus.APPROVED,
            transaction_id=transaction_id,
        )

        print("[SUCCESS] Payment processed successfully!")
        print(f"[DEBUG] Transaction ID: {transaction_id}")

        print(f"[LOG] Payment completed - Card: ****{payment_info.card_last_four_digits}, Amount: ${amount}, Transaction: {transaction_id}")

        return True

    # Returns only the last 4 digits of the card number for safe display and logging
    @staticmethod
    def _mask_card_number(card_number: str) -> str:
        digits_only = "".join(c for c in (card_number or "") if c.isdigit())
        return digits_only[-4:]

    # Detects the card brand from the Issuer Identification Number range
    @staticmethod
    def _detect_card_type(card_number: str) -> str:
        digits_only = "".join(c for c in (card_number or "") if c.isdigit())

        if digits_only.startswith("4"):
            return "Visa"
        if len(digits_only) >= 2 and 51 <= int(digits_only[:2]) <= 55:
            return "Mastercard"
        if digits_only.startswith(("34", "37")):
            return "American Express"
        if digits_only.startswith(("6011", "65")):
            return "Discover"
        return "Unknown"

    # Vulnerable card validation
    def _validate_card_number(self, card_number: str) -> bool:
        # Security vulnerability: Weak validation - only checks length
        if not card_number:
            return False

        # Remove spaces and dashes
        card_number = card_number.replace(" ", "").replace("-", "")

        # Security vulnerability: Accept any 13-19 digit number
        return 13 <= len(card_number) <= 19 and card_number.isdigit()

    def _validate_expiry_date(self, expiry_date: str) -> bool:
        # Security vulnerability: Basic validation only
        if not expiry_date or "/" not in expiry_date:
            return False

        parts = expiry_date.split("/")
        if len(parts) != 2 or not parts[0].isdigit() or not parts[1].isdigit():
            return False

        month, year = int(parts[0]), int(parts[1])
        if year < 100:
            year += 2000  # Convert YY to YYYY
        expiry = datetime(year, month, monthrange(year, month)[1])
        return expiry >= datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)

    # Security vulnerability: Predictable transaction ID generation
    def _generate_transaction_id(self, card_number: str, amount: Decimal) -> str:
        # Vulnerable: Using predictable pattern
        last_four = card_number[-4:]
        timestamp = datetime.now().strftime("%Y%m%d%H%M")
        amount_str = f"{amount:.2f}".replace(".", "")

        return f"TXN_{timestamp}_{last_four}_{amount_str}"

    def refund_payment(self, transaction_id: str, amount: Decimal) -> bool:
        # Security vulnerability: Log refund details
        print(f"[DEBUG] Processing refund for transaction: {transaction_id}, Amount: ${amount}")
        print(f"[DEBUG] Using payment gateway: {self.PAYMENT_GATEWAY_URL}")

        # Simulate refund processing
        print("[INFO] Processing refund...")
        time.sleep(0.5)

        print(f"[SUCCESS] Refund processed for transaction: {transaction_id}")
        return True

    # Method to get payment history - with security issues
    def get_payment_history(self, user_id: int) -> list[PaymentInfo]:
        print(f"[DEBUG] Retrieving payment history for user: {user_id}")

        # In a real app, this would query the database
        # For demo purposes, we'll return an empty list
        return []
